src/resources/ble_esp32.cc

```diff
@@ -460,8 +458,13 @@ class BleServiceResource:
   BlePeripheralManagerResource* peripheral_manager() const { return peripheral_manager_; }
   CharacteristicResourceList& characteristics() { return characteristics_; }
 
+  void remove_characteristic(BleCharacteristicResource* characteristic) {
+    characteristics_.unlink(characteristic);
+  }
+
   void clear_characteristics() {
-    while (auto characteristic = characteristics_.remove_first()) {
+    while (!characteristics_.is_empty()) {
+      auto characteristic = characteristics_.first();
       group()->unregister_resource(characteristic);
```

Comment:
Maybe add a comment that explains that unregistering will remove the element from linked list?

Reply:
done.